substrate/frame/delegated-staking/src/tests.rs

```diff
@@ -0,0 +1,647 @@
+			DelegatedStaking::register_agent(
+				RawOrigin::Signed(mock::GENESIS_NOMINATOR_ONE).into(),
+				100
+			),
+			Error::<T>::AlreadyStaking
+		);
+		assert_noop!(
+			DelegatedStaking::register_agent(
+				RawOrigin::Signed(mock::GENESIS_NOMINATOR_TWO).into(),
+				100
+			),
+			Error::<T>::AlreadyStaking
+		);
+	});
+}
+
+#[test]
+fn create_multiple_delegators() {
+	ExtBuilder::default().build_and_execute(|| {
+		let agent: AccountId = 200;
+		let reward_account: AccountId = 201;
+
+		// stakeable balance is 0 for non agent
+		fund(&agent, 1000);
+		assert!(!DelegatedStaking::is_agent(&agent));
+		assert_eq!(DelegatedStaking::stakeable_balance(&agent), 0);
+
+		// set intention to accept delegation.
+		assert_ok!(DelegatedStaking::register_agent(
+			RawOrigin::Signed(agent).into(),
+			reward_account
+		));
+
+		// create 100 delegators
+		for i in 202..302 {
+			fund(&i, 100 + ExistentialDeposit::get());
+			assert_ok!(DelegatedStaking::delegate_to_agent(
+				RawOrigin::Signed(i).into(),
+				agent,
+				100
+			));
+			// Balance of 100 held on delegator account for delegating to the agent.
+			assert_eq!(Balances::balance_on_hold(&HoldReason::StakingDelegation.into(), &i), 100);
+		}
+
+		// verify
+		assert!(DelegatedStaking::is_agent(&agent));
+		assert_eq!(DelegatedStaking::stakeable_balance(&agent), 100 * 100);
+	});
+}
+
+#[test]
+fn agent_restrictions() {
+	// Similar to creating a nomination pool
+	ExtBuilder::default().build_and_execute(|| {
+		let agent_one = 200;
+		let delegator_one = 210;
+		fund(&agent_one, 100);
+		assert_ok!(DelegatedStaking::register_agent(
+			RawOrigin::Signed(agent_one).into(),
+			agent_one + 1
+		));
+		fund(&delegator_one, 200);
+		assert_ok!(DelegatedStaking::delegate_to_agent(
+			RawOrigin::Signed(delegator_one).into(),
+			agent_one,
+			100
+		));
+
+		let agent_two = 300;
+		let delegator_two = 310;
+		fund(&agent_two, 100);
+		assert_ok!(DelegatedStaking::register_agent(
+			RawOrigin::Signed(agent_two).into(),
+			agent_two + 1
+		));
+		fund(&delegator_two, 200);
+		assert_ok!(DelegatedStaking::delegate_to_agent(
+			RawOrigin::Signed(delegator_two).into(),
+			agent_two,
+			100
+		));
+
+		// agent one tries to delegate to agent 2
+		assert_noop!(
+			DelegatedStaking::delegate_to_agent(RawOrigin::Signed(agent_one).into(), agent_two, 10),
+			Error::<T>::InvalidDelegation
+		);
+
+		// agent one tries to delegate to a delegator
+		assert_noop!(
+			DelegatedStaking::delegate_to_agent(
+				RawOrigin::Signed(agent_one).into(),
+				delegator_one,
+				10
+			),
+			Error::<T>::InvalidDelegation
+		);
+		assert_noop!(
+			DelegatedStaking::delegate_to_agent(
+				RawOrigin::Signed(agent_one).into(),
+				delegator_two,
+				10
+			),
+			Error::<T>::InvalidDelegation
+		);
+
+		// delegator one tries to delegate to agent 2 as well (it already delegates to agent
+		// 1)
+		assert_noop!(
+			DelegatedStaking::delegate_to_agent(
+				RawOrigin::Signed(delegator_one).into(),
+				agent_two,
```

Comment:
Would expect to see `fn delegator_restrictions` that ensures existing delegators:

1. cannot delegate to others
2. cannot delegate to non-agents.

Reply:
Not sure what you mean by fn `delegator_restrictions` but added new assertions to ensure
- delegator cannot delegate to non-agents.
- delegator cannot delegate to other delegators/self.
- agents cannot delegate to other agents or self.

> cannot delegate to others
there were already existing assertions for this.